kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/empty.kt

```diff
@@ -7,50 +7,140 @@ import io.kotest.matchers.invokeMatcher
 import io.kotest.matchers.should
 import io.kotest.matchers.shouldNot
 
-fun <T, I : Iterable<T>> I?.shouldBeEmpty(): I {
-   if (this == null) fail()
-   toList().shouldBeEmpty()
+fun BooleanArray?.shouldBeEmpty(): BooleanArray {
+   if (this == null) fail("BooleanArray")
+   asList() should beEmpty("BooleanArray")
    return this
 }
 
-fun <T> Array<T>?.shouldBeEmpty(): Array<T> {
-   if (this == null) fail()
-   this should beEmptyArray()
+fun BooleanArray?.shouldNotBeEmpty(): BooleanArray {
+   if (this == null) fail("BooleanArray")
+   asList() shouldNot beEmpty("BooleanArray")
    return this
 }
 
-fun <T, C : Collection<T>> C?.shouldBeEmpty(): C {
-   if (this == null) fail()
-   this should beEmpty()
+fun ByteArray?.shouldBeEmpty(): ByteArray {
+   if (this == null) fail("ByteArray")
+   asList() should beEmpty("ByteArray")
    return this
 }
 
-fun <T, I : Iterable<T>> I?.shouldNotBeEmpty(): I {
-   if (this == null) fail()
-   toList().shouldNotBeEmpty()
+fun ByteArray?.shouldNotBeEmpty(): ByteArray {
+   if (this == null) fail("ByteArray")
+   asList() shouldNot beEmpty("ByteArray")
+   return this
+}
+
+fun ShortArray?.shouldBeEmpty(): ShortArray {
+   if (this == null) fail("ShortArray")
+   asList() should beEmpty("ShortArray")
+   return this
+}
+
+fun ShortArray?.shouldNotBeEmpty(): ShortArray {
+   if (this == null) fail("ShortArray")
+   asList() shouldNot beEmpty("ShortArray")
+   return this
+}
+
+fun CharArray?.shouldBeEmpty(): CharArray {
+   if (this == null) fail("CharArray")
+   asList() should beEmpty("CharArray")
+   return this
+}
+
+fun CharArray?.shouldNotBeEmpty(): CharArray {
+   if (this == null) fail("CharArray")
+   asList() shouldNot beEmpty("CharArray")
+   return this
+}
+
+fun IntArray?.shouldBeEmpty(): IntArray {
+   if (this == null) fail("IntArray")
+   asList() should beEmpty("IntArray")
+   return this
+}
+
+fun IntArray?.shouldNotBeEmpty(): IntArray {
+   if (this == null) fail("IntArray")
+   asList() shouldNot beEmpty("IntArray")
+   return this
+}
+
+fun LongArray?.shouldBeEmpty(): LongArray {
+   if (this == null) fail("LongArray")
+   asList() should beEmpty("LongArray")
+   return this
+}
+
+fun LongArray?.shouldNotBeEmpty(): LongArray {
+   if (this == null) fail("LongArray")
+   asList() shouldNot beEmpty("LongArray")
+   return this
+}
+
+fun FloatArray?.shouldBeEmpty(): FloatArray {
+   if (this == null) fail("FloatArray")
+   asList() should beEmpty("FloatArray")
+   return this
+}
```

Comment:
I don't think there is any reasonable way to change this since `FloatArray != Array<Float>`. There is no common supertype for primitive arrays. If you wanted a single generic function `shouldBeEmpty` that works across all primitive arrays, you would have to make it an extension on `T` - and that is not desirable IMO.

Reply:
Right now I believe a reasonable balance has been struck - all significant logic is extracted to a single matcher and the assertions which use it are organized in a way that can easily be checked by human eye for correctness.